Wrappers/Python/cil/optimisation/functions/TotalGeneralisedVariation.py

```diff
@@ -0,0 +1,211 @@
+
+
+        Parameters
+        ----------
+        max_iteration : :obj:`int`, default = 100
+            Maximum number of iterations for the PDHG algorithm.
+        correlation : :obj:`str`, default = `Space`
+            Correlation between `Space` and/or `SpaceChannels` for the :class:`.GradientOperator`.
+        backend :  :obj:`str`, default = `c`      
+            Backend to compute the :class:`.GradientOperator`  
+        split : :obj:`boolean`, default = False
+            Splits the Gradient into spatial gradient and spectral or temporal gradient for multichannel data.
+        verbose : :obj:`int`, default = 0
+            Prints primal, dual and primal-dual objectives of the :class:`PDHG` algorithm.
+        update_objective_interval : :obj:`int`, default = :code:`max_iteration`
+            Updates the objectives of the :class:`PDHG` algorithm.
+
+
+        Examples
+        --------
+
+        .. math:: \mathrm{prox}_{\tau\mathrm{TGV}_{\alpha, \beta}}(b) = \underset{u}{\mathrm{argmin}} \frac{1}{2\tau}\|u - b\|^{2} + \mathrm{TGV}_{\alpha, \beta}(u) 
+
+
+        >>> alpha1 = 1.0
+        >>> alpha0 = 2.0
+        >>> TGV = TotalGeneralisedVariation(alpha1 = alpha1, alpha0 = alpha0)
+        >>> sol = TGV.proximal(b, tau = 1.0)
+
+
+        """   
+        
+            
+    def __init__(self,
+                 reg_parameter=1.0,
+                 alpha1 = 1.0,
+                 alpha0 = 2.0,
+                 max_iteration = 100, 
+                 update_objective_interval = 50,
+                 correlation = "Space",
+                 backend = "c",
+                 split = False,
+                 verbose = 0,                  
+                 warmstart=False):
+        
+        super(TotalGeneralisedVariation, self).__init__(L = None)
+
+        # regularisation parameters for TGV
+        self.alpha1 = alpha1
+        self.alpha0 = alpha0
+        self.reg_parameter = reg_parameter
+                
+        # Iterations for PDHG_TGV
+        self.iterations = max_iteration
+        
+        # correlation space or spacechannels
+        self.correlation = correlation
+
+        # backend for the gradient
+        self.backend = backend        
+        
+        # splitting Gradient
+        self.split = split
+                        
+        self.verbose = verbose
+        self.update_objective_interval = update_objective_interval
+
+        # warm-start
+        self.warmstart  = warmstart
+        if self.warmstart:
+            self.hasstarted = False        
+
+
+    def __call__(self, x):
+        
+        if not hasattr(self, 'pdhg'):   
+            return 0.0        
+        else:              
+            # Compute alpha1 * || Du - w || + alpha0 * ||Ew||, 
+            # where (u,w) are solutions coming from the proximal method below.
+            tmp = self.f(self.pdhg.operator.direct(self.pdhg.solution))
+            return tmp
+
+    def _setup_solver(self, x, tau=1.0):
```

Comment:
This method is called `_setup_solver` but it actually also runs it. Maybe `_setup_and_run_solver` would be better.